---
title: "SingleR Tabula"
author: "Andrew Liu"
date: "2023-06-16"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(Seurat)
library(patchwork)
library(tidyverse)
library(readr)
library(ggplot2)
library(celldex) #Like Rolodex for cell datasets; all functions call datasets
library(scRNAseq) #Single Cell RNA sequence datasets
library(SingleR) #Use to bring in SingleR() function, which compares reference and test data for gene similarity
library(scuttle) #logNormCounts function
library(biomaRt)
```

```{r}
TS_panc <- readRDS(file = "TS_Pancreas.rds")
#TS_panc <- UpdateSeuratObject(TM_panc)
TS_panc <- as.SingleCellExperiment(TS_panc)
TS_panc <- logNormCounts(TS_panc)

Mur_panc <- MuraroPancreasData()
Mur_panc <- Mur_panc[, !is.na(Mur_panc$label)]
#Mur_panc$label[which(is.na(Mur_panc))] <- "Unclassified"
Mur_panc <- logNormCounts(Mur_panc)

rownames(Mur_panc) = gsub("_.*","",rownames(Mur_panc))

ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
listAttributes(ensembl)

gene_ids <- getBM(attributes = c("ensembl_gene_id", "external_gene_name"),
                  filters = "ensembl_gene_id",
                  values = TS_panc@rowRanges@partitioning@NAMES,
                  mart = ensembl)

gene_e_ids <- gene_ids$ensembl_gene_id
gene_id_names <- gene_ids$external_gene_name

TS_e_ids <- TS_panc@rowRanges@partitioning@NAMES

gene_id_to_name <- setNames(gene_id_names, gene_e_ids)

matched_gene_names <- gene_id_to_name[TS_e_ids]
TS_panc@rowRanges@partitioning@NAMES <- matched_gene_names

#nameVect = vector()
#origIdent = TM_panc@rowRanges@partitioning@NAMES
#nameVect = gene_ids$external_gene_name[match(gene_ids$ensembl_gene_id,origIdent)]

#TM_panc@rowRanges@partitioning@NAMES <- TM_panc@rowRanges@partitioning@NAMES %>%
 # group_by()

pred.TS <- SingleR(test = Mur_panc,
                   ref = TS_panc,
                   assay.type.test = 1,
                   labels = TS_panc$cell_ontology_class)
table(pred.TS$labels)
#Labels are used from the REFERENCE dataset
#Gene abbreviation: A1BG
#Gene ID: numeric IDs
#Ensemble IDs: identifiers from the comprehensive ensembl dataset (like a social security number for genes)
```
```{r}
plotScoreHeatmap(pred.TS)
plotDeltaDistribution(pred.TS, ncol = 3)
summary(is.na(pred.TS$pruned.labels))
```

```{r}
muraro.seurat <- MuraroPancreasData()
muraro.seurat <- logNormCounts(muraro.seurat)

muraro.seurat@rowRanges@partitioning@NAMES <- make.unique(rownames(muraro.seurat))

muraro.seurat <- as.Seurat(muraro.seurat, counts = "counts", data = NULL)
muraro.seurat <- NormalizeData(muraro.seurat)
DefaultAssay(muraro.seurat) <- "endogenous"
#Currently, muraro.seurat shows both endogenous data and ERCC data.
```

```{r}
saveRDS(pred.TS, "tabulaSapiens_muraro.rds")
#muraro.seurat[["TabulaSapiensPrediction"]] = pred.TS$pruned.labels

tabulaSapiens = vector(length = ncol(muraro.seurat))
names(tabulaSapiens) = colnames(muraro.seurat)
tabulaSapiens[rownames(pred.TS)] = pred.TS[rownames(pred.TS), "pruned.labels"]
#We define the tabulaSapiens vector as having column names taken from muraro.seurat. We then index tabulaSapiens, and for each colnames (from muraro.seurat) that matches up with a rowname from the TS prediction, we assign the pruned.label value to that index in the vector. We end up with a table full of different cell types (labels).
table(tabulaSapiens)
muraro.seurat[["TabulaSapiensPrediction"]] = tabulaSapiens
#We then assign the tabulaSapiens vector into the TabulaSapiensPrediction column.
table(muraro.seurat$TabulaSapiensPrediction, muraro.seurat$label)
#Compare predictions with actual labels.

muraro.seurat = ScaleData(muraro.seurat)
#Log transform, center, and scale data in one step for downstream analysis.
muraro.seurat = FindVariableFeatures(muraro.seurat)
#
muraro.seurat = RunPCA(muraro.seurat, npcs=50)

muraro.seurat = FindNeighbors(muraro.seurat, dims=1:20)

muraro.seurat = RunUMAP(muraro.seurat,dims=1:20)

scatter1 <- DimPlot(muraro.seurat, group.by = "TabulaSapiensPrediction", reduction = "umap", label = TRUE, label.size = 2)
scatter2 <- DimPlot(muraro.seurat, group.by = "label", reduction = "umap", label = TRUE, label.size = 2)

scatter1 + scatter2
scatter1
scatter2
#muraro.seurat[["percent.mt"]] <- PercentageFeatureSet(muraro.seurat, pattern = "^MT-")
#grep("^MT-",rownames(muraro.seurat),value=T)

#{muraro.seurat[["percent.mt"]] <- PercentageFeatureSet(muraro.seurat, pattern = "^MT-")}

#if(species=="hg38"){so[["percent.mt"]] <- PercentageFeatureSet(so, pattern = "^MT-")}
#if(species="mm10"){so[["percent.mt"]] <- PercentageFeatureSet(so, pattern = "^mt-")}

#head(muraro.seurat@meta.data, 50)
#VlnPlot(muraro.seurat, features = c("nFeature_endogenous", "nCount_endogenous", "percent.mt"), ncol = 3)
```